x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -245,55 +256,107 @@ public void clusterStateProcessed(ClusterState oldState, ClusterState newState)
+                    .removeCustom(TrainedModelAssignmentMetadata.DEPRECATED_NAME)
+            )
+            .build();
     }
 
+    /** Visible for testing */
     ClusterState createModelAssignment(ClusterState currentState, StartTrainedModelDeploymentAction.TaskParams params) {
-        if (MlMetadata.getMlMetadata(currentState).isResetMode()) {
-            throw new ElasticsearchStatusException(
-                "cannot create new assignment for model [{}] while feature reset is in progress.",
-                RestStatus.CONFLICT,
-                params.getModelId()
-            );
-        }
-        TrainedModelAssignmentMetadata.Builder builder = TrainedModelAssignmentMetadata.builder(currentState);
-        if (builder.hasModel(params.getModelId())) {
-            throw new ResourceAlreadyExistsException("assignment for model with id [{}] already exist", params.getModelId());
-        }
-        TrainedModelAssignment.Builder assignmentBuilder = TrainedModelAssignment.Builder.empty(params);
-
-        Set<String> shuttingDownNodes = nodesShuttingDown(currentState);
-        Map<String, String> nodeToReason = new TreeMap<>();
-        for (DiscoveryNode node : currentState.getNodes()) {
-            if (StartTrainedModelDeploymentAction.TaskParams.mayAssignToNode(node) && shuttingDownNodes.contains(node.getId()) == false) {
-                Optional<String> maybeError = nodeHasCapacity(currentState, params, node);
-                if (maybeError.isPresent()) {
-                    nodeToReason.put(node.getName(), maybeError.get());
-                } else {
-                    assignmentBuilder.addNewRoutingEntry(node.getId());
-                }
+        return update(currentState, rebalanceAssignments(currentState, Optional.of(params)));
+    }
+
+    private void rebalanceAssignments(
+        ClusterState clusterState,
+        Optional<StartTrainedModelDeploymentAction.TaskParams> modelToAdd,
+        String reason,
+        ActionListener<TrainedModelAssignmentMetadata> listener
+    ) {
+        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+            logger.debug(() -> format("Rebalancing model allocations because [%s]", reason));
+            TrainedModelAssignmentMetadata.Builder rebalancedMetadata;
+            try {
+                rebalancedMetadata = rebalanceAssignments(clusterState, modelToAdd);
+            } catch (Exception e) {
+                listener.onFailure(e);
+                return;
             }
-        }
-        if (nodeToReason.isEmpty() == false) {
-            assignmentBuilder.setReason(
-                nodeToReason.entrySet()
-                    .stream()
-                    .map(entry -> String.format(Locale.ROOT, "Not allocating on node [%s]. Reason: %s", entry.getKey(), entry.getValue()))
-                    .collect(Collectors.joining("|"))
+            submitUnbatchedTask(reason, new ClusterStateUpdateTask() {
```

Comment:
We should totally allow these assignments updates to be batched. We can at least keep track of the individual creation requests and handle them all at once in a batched manner. This would require creating these cluster update task objects, keeping track of the listeners, and putting them together, but we should do this extra work.

This will at least reduce the amount of work required by the "rebalanceAssignments" as we know we have all the assignments requested at once. Just not anomaly detection jobs.

Reply:
https://github.com/elastic/elasticsearch/issues/87551